### PR TITLE
Improvements to My Site tab.

### PR DESCRIPTION
**Pull request by diegoreymendez**
Swaps the list of sites and blog details when the `newNavBarAppearance` flag is on.

Should not change anything when the flag is off.

## Known limitations:

When the `newNavBarAppearance` feature flag is ON there's currently no way to switch sites.  I've disabled it temporarily until I can implement the site switching.

## To test:

The most important testing that can be done is when the feature flag `newNavBarAppearance` is off, since that's the experience that our users will have in the release version of the App.

Please take a moment to thoroughly test and review the code in this PR.

### Test 1:

Feature flag `newNavBarAppearance` ON.
Use an account that has at least 1 blog.

1. Tap "My Site" in the tab bar
2. Make sure the first VC shown the blog details for your last used blog, or your main blog.
3. Make sure there's no option to go to the list of sites just yet (coming up in my next PR).

### Test 2:

Feature flag `newNavBarAppearance` ON.
Use an account that has no blogs.

1. Tap "My Site" in the tab bar
2. Make sure you see a VC that says you have no blogs.
3. Create a site, make sure once you're done you see the blog details for the blog you created.
4. Delete the site, make sure once you're done you see a no-results view stating you have no blogs.

### Test 3:

Feature flag `newNavBarAppearance` OFF.
Use an account that has at least 1 blog.

1. Tap "My Site" in the tab bar
2. Make sure the first VC shown the blog details for your last used blog, or your main blog.
3. Make sure you can tap on the back button in the navigation bar to pick a different site.
4. Switch to another site and make sure it works fine.

### Test 4:

Feature flag `newNavBarAppearance` OFF.
Use an account that has no blogs.

1. Tap "My Site" in the tab bar
2. Make sure the first VC shown is a no-results saying you have no sites.
3. Create a site.
4. Make sure you see the list of blogs with the site you created.
5. Delete the site.
6. Make sure you see the no results view again.

### Additional notes:

In all tests, make sure you also tap in the user's icon, and that the Me VC comes up just fine.

## PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m [hunk 1]
```diff
@@ -251,18 +251,8 @@ + (UIViewController *)viewControllerWithRestorationIdentifierPath:(NSArray *)ide
...
-    // navigation stack, we'll return that instead of creating a new one.
-    UISplitViewController *splitViewController = [[WPTabBarController sharedInstance] blogListSplitViewController];
-    UINavigationController *navigationController = splitViewController.viewControllers.firstObject;
-    if (navigationController && [navigationController isKindOfClass:[UINavigationController class]]) {
-        BlogDetailsViewController *topViewController = (BlogDetailsViewController *)navigationController.topViewController;
-        if ([topViewController isKindOfClass:[BlogDetailsViewController class]] && topViewController.blog == restoredBlog) {
-            return topViewController;
-        }
-    }
+    BlogDetailsViewController *viewController = [[BlogDetailsViewController alloc] initWithMeScenePresenter:[MeScenePresenter new]];
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m:256]: Is there a reason we're no longer checking if there's already an existing VC we can reuse?

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            guard let newBlog = newValue else {
+                showBlogDetailsForMainBlogOrNoSites()
+                return
+            }
+
+            showBlogDetails(for: newBlog)
+        }
+
+        get {
+            guard let blogDetailsViewController = blogDetailsViewController else {
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Can we not just return `blogDetailsViewController?.blog` ? (will be nil if there's no blogDetailsViewController)

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            showNoSites()
+            return
+        }
+
+        showBlogDetails(for: mainBlog)
+    }
+
+    // MARK: - No Sites UI logic
+
+    private func hideNoSites() {
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:117]: I can't see that this is ever called anywhere?

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    /// In order to adhere to this logic, if this VC is shown without a blog being set, we will try to load the "main" blog (ie in order: the last used blog,
+    /// the account's primary blog, or the first blog we find for the account).
+    ///
+    private func showBlogDetailsForMainBlogOrNoSites() {
+        guard let mainBlog = mainBlog() else {
+            showNoSites()
+            return
+        }
+
+        showBlogDetails(for: mainBlog)
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:112]: Should this method or `showBlogDetails(for:)` hide the no results view if it exists?

File: WordPress/Classes/ViewRelated/System/WPTabBarController.m [hunk 5]
```diff
@@ -371,14 +319,13 @@ - (ReaderCoordinator *)readerCoordinator
 
 - (NSArray<UIViewController *> *)tabViewControllers
 {
-    
     NSMutableArray<UIViewController *> *allViewControllers;
     if ([Feature enabled:FeatureFlagNewReaderNavigation]) {
-        allViewControllers = [NSMutableArray arrayWithArray:@[self.blogListSplitViewController,
+        allViewControllers = [NSMutableArray arrayWithArray:@[self.mySitesCoordinator.rootViewController,
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/System/WPTabBarController.m]: I'm not sure either of these return arrays need to be mutable

File: WordPress/Classes/ViewRelated/System/WPTabBarController.m [hunk 6]
```diff
@@ -610,7 +505,7 @@ - (BOOL)tabBarController:(UITabBarController *)tabBarController shouldSelectView
     if (selectedIndex != tabBarController.selectedIndex) {
         switch (selectedIndex) {
             case WPTabMySites: {
-                [self bypassBlogListViewControllerIfNecessary];
+                [self.mySitesCoordinator showMainViewControllerWithCanBypassBlogList:true];
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/System/WPTabBarController.m]: `true` -> `YES`

File: WordPress/Classes/ViewRelated/System/WPTabBarController.m [hunk 7]
```diff
@@ -659,7 +539,7 @@ - (void)showNotificationsTabForNoteWithID:(NSString *)notificationID
 
 - (BOOL)isNavigatingMySitesTab
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/System/WPTabBarController.m]: Looks like this method might not be used at all

**Review by frosty — COMMENTED** [review]
This is looking really good in general – I like the use of the MySitesCoordinator to encapsulate a bunch of this logic, and the new MySiteViewController. I ran through your testing scenarios and everything was mostly fine. I've added a few notes to the code, and I spotted a few other issues, however:

- Tapping the Add New Site button in the no results view doesn't do anything when the feature flag is enabled (not such a big deal, as long as we fix it later!)
- Blog details screen isn't always scrolled to the top when displayed
- The biggest issue: I was logged in with an account with many sites, then logged out and back in with an account with no sites. Got a crash:
```
2020-11-12 19:35:25:879 WordPress[49667:2444564] Blog display URL is nil
2020-11-12 19:35:25.880999+0000 WordPress[49667:2444564] *** Assertion failure in -[AccountService initWithManagedObjectContext:], LocalCoreDataService.m:11
2020-11-12 19:35:25.909822+0000 WordPress[49667:2444564] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid parameter not satisfying: [context isKindOfClass:[NSManagedObjectContext class]]'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff2043a126 __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007fff20177f78 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff20439f4f +[NSException raise:format:] + 0
	3   Foundation                          0x00007fff207881ca -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 191
	4   WordPress                           0x000000010cc21f9f -[LocalCoreDataService initWithManagedObjectContext:] + 255
	5   WordPress                           0x000000010cc942ba -[Blog accountIsDefaultAccount] + 90
	6   WordPress                           0x000000010cc93811 -[Blog supports:] + 81
	7   WordPress                           0x000000010ccda9ad -[BlogDetailsViewController configureTableViewData] + 1133
	8   WordPress                           0x000000010ccd7d67 -[BlogDetailsViewController viewWillAppear:] + 903
	9   UIKitCore                           0x00007fff23f4265a -[UIViewController _setViewAppearState:isAnimating:] + 654
	10  UIKitCore                           0x00007fff23f42cdd __52-[UIViewController _setViewAppearState:isAnimating:]_block_invoke + 276
	11  CoreFoundation                      0x00007fff203ab85b __NSARRAY_IS_CALLING_OUT_TO_A_BLOCK__ + 7
	12  CoreFoundation                      0x00007fff20482e6f -[__NSSingleObjectArrayI enumerateObjectsWithOptions:usingBlock:] + 80
	13  UIKitCore                           0x00007fff23f42941 -[UIViewController _setViewAppearState:isAnimating:] + 1397
	14  UIKitCore                           0x00007fff23f42e83 -[UIViewController __viewWillAppear:] + 106
	15  UIKitCore                           0x00007fff23e67f14 -[UINavigationController viewWillAppear:] + 432
	16  UIKitCore                           0x00007fff23f4265a -[UIViewController _setViewAppearState:isAnimating:] + 654
	17  UIKitCore                           0x00007fff23f42cdd __52-[UIViewController _setViewAppearState:isAnimating:]_block_invoke + 276
	18  CoreFoundation                      0x00007fff203ab85b __NSARRAY_IS_CALLING_OUT_TO_A_BLOCK__ + 7
	19  CoreFoundation                      0x00007fff2045b75c -[__NSArrayI enumerateObjectsWithOptions:usingBlock:] + 176
	20  UIKitCore                           0x00007fff23f42941 -[UIViewController _setViewAppearState:isAnimating:] + 1397
	21  UIKitCore                           0x00007fff23f42e83 -[UIViewController __viewWillAppear:] + 106
	22  UIKitCore                           0x00007fff23e3c090 -[UITabBarController viewWillAppear:] + 293
	23  UIKitCore                           0x00007fff23f4265a -[UIViewController _setViewAppearState:isAnimating:] + 654
	24  UIKitCore                           0x00007fff23f42e83 -[UIViewController __viewWillAppear:] + 106
	25  UIKitCore                           0x00007fff23e1b01e __56-[UIPresentationController runTransitionForCurrentState]_block_invoke.466 + 1201
	26  UIKitCore                           0x00007fff24b66632 -[_UIAfterCACommitBlock run] + 54
	27  UIKitCore                           0x00007fff2467c179 _runAfterCACommitDeferredBlocks + 333
	28  UIKitCore                           0x00007fff2466c0ac _cleanUpAfterCAFlushAndRunDeferredBlocks + 221
	29  UIKitCore                           0x00007fff2469da70 _afterCACommitHandler + 85
	30  CoreFoundation                      0x00007fff203a76b3 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	31  CoreFoundation                      0x00007fff203a1f3f __CFRunLoopDoObservers + 547
	32  CoreFoundation                      0x00007fff203a24e2 __CFRunLoopRun + 1113
	33  CoreFoundation                      0x00007fff203a1b9e CFRunLoopRunSpecific + 567
	34  GraphicsServices                    0x00007fff2b793db3 GSEventRunModal + 139
	35  UIKitCore                           0x00007fff2466d40f -[UIApplication _run] + 912
	36  UIKitCore                           0x00007fff24672320 UIApplicationMain + 101
	37  WordPress                           0x000000010db4de74 main + 500
	38  libdyld.dylib                       0x00007fff20257409 start + 1
	39  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid parameter not satisfying: [context isKindOfClass:[NSManagedObjectContext class]]'
terminating with uncaught exception of type NSException
```

This happened on iPad. I just tested it on iPhone and couldn't replicate it, but I will try again and let you know.

**Comment by frosty** [comment]
@diegoreymendez FYI I just created the crash again, this time on iPhone. Steps were:

- Either fresh install, or log out of the app and quit it
- Run the app and log in with account with multiple sites
- Log out
- Log in with account with no sites
- Tap the done button at the end of the login prologue
- Crash

**Comment by Gio2018** [comment]
> @diegoreymendez FYI I just created the crash again, this time on iPhone. Steps were:
> 
> * Either fresh install, or log out of the app and quit it
> * Run the app and log in with account with multiple sites
> * Log out
> * Log in with account with no sites
> * Tap the done button at the end of the login prologue
> * Crash

Tried this one on device iPhone X iOS 14.1 and crashed as well

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 8]
```diff
@@ -0,0 +1,255 @@
+import CoreData
```
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Maybe better to do in a separate PR, but would it be possible to abstract away anything that is `CoreData` related?

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+import Foundation
```
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Nit: if `UIKit` is imported by default, we probably don't need this

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    // MARK: - Initializers
+
+    init(meScenePresenter: ScenePresenter) {
+        self.meScenePresenter = meScenePresenter
+
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        meScenePresenter = MeScenePresenter()
```
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: I understand this if we want to still allow to instantiate this via storyboard, but if for any reason we substitute the `ScenePresenter` concrete type, this will need to be handled manually

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+            }
+
+            return blogDetailsViewController.blog
+        }
+    }
+
+    /// The VC for the blog details.  This class is written in a way that this VC will only exist if it's being shown on screen.
+    /// Please keep this in mind when making modifications.
+    ///
+    private var blogDetailsViewController: BlogDetailsViewController? = nil
```
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Nit: explicitly initializing to `nil` should not be needed

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+            self.blogDetailsViewController = blogDetailsViewController
+            return blogDetailsViewController
+        }
+
+        blogDetailsViewController.blog = blog
+        return blogDetailsViewController
+    }
+
+    private func makeBlogDetailsViewController(for blog: Blog) -> BlogDetailsViewController {
+        let blogDetailsViewController = BlogDetailsViewController(meScenePresenter: MeScenePresenter())
```
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: we are not using all the benefits of dependency injection here, since we are directly instantiating one concrete type. We could use some sort of `factory` that this class uses to instantiate the presenter, if we want to abstract away this, but it might be overkilling.

**Review by Gio2018 — COMMENTED** [review]
Agree with @frosty 's comments and added a few more myself. I tried the steps to repro the crash and it happened on iPhone X 14.1

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 12 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Once the feature flag is removed, `BlogDetailsViewController` will / should only be instantiated / shown by `MySiteViewController`.  It could even be made private to this class.

Due to this, I don't think there's a need to abstract its instantiation, as it should not be used elsewhere, and changing the VC internally for `MySiteViewController` would be a one-liner.

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 10 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: I could make this fatally fail if you prefer.  I'm ok either way since this initialiser isn't used.

Let me know which one feels best.

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 8 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Well, regarding the abstraction everything can be discussed.  But this header isn't really needed so I went ahead and removed it.

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 3 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:117]: I've added the call following your suggestion in the next comment.

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 10 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: we could make it fail

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 2 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Good call, dunno why I missed this.

File: WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m (hunk 1 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m:256]: Honestly I couldn't trigger this code in all of my tests.  Let's keep this conversation open and I'll try to figure out tomorrow if this code is being used and either remove it, or make sure it's fine.

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 12 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Sounds good, I guess we can still use the `ScenePresenter` that is injected in `MySiteViewController`?

**Comment by diegoreymendez** [comment]
Thanks for the thorough review @frosty and @Gio2018 .  I've addressed the crash (plus some other crash that felt unrelated, but was surfaced by my changes).  I've also addressed most feedback.

So to summarize:

- Most feedback I've addressed, including the crasher.  Feel free to validate and retest all resolved ones.
- Some feedack I've asked for more details / ideas.
- The state restoration thing pointed by @frosty I'll be testing tomorrow.

Can I ask you to retest the fixed issues, and check my comments on the other ones?

Thanks!

**Comment by diegoreymendez** [comment]
Oh additionally...

I've added the code to add new sites / self hosted sites, so that's testable.

For context on why this was missing - I had left it in my old / discontinued draft PR and forgot to add it to this one: https://github.com/wordpress-mobile/WordPress-iOS/pull/15130

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 12 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Apologies, I misunderstood the feedback.  You're right.

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift (hunk 8 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift]: Marking as resolved for now but feel free to respond if there's anything else.

**Review by Gio2018 — CHANGES_REQUESTED** [review]
Hey @diegoreymendez , in dark mode, the header seems to have the wrong color (thank you @emilylaguna for spotting this!):

<p align=center>

</p>

also, with the feature flag turned off and an account with at least one site, if I tap on the reader and then back to My Site, it goes on the list of sites instead of the previously selected site

<p align=center>

</p>

finally, dunno if that's expected, but the white border of the "Me" avatar does not show when the flag is turned on, and the titles slightly differ ("My Site" vs "My Sites")

flag on | flag off
-- | --
 |

**Comment by diegoreymendez** [comment]
Thanks @Gio2018... you saved me a lot of trouble with that.  I'll change this PR back to draft and resubmit once I've had a chance to check that out.

**Comment by jkmassel** [comment]
👋 Howdy! We are freezing `16.2` today, so this PR is being pushed to `16.3`. If you need it to be part of the `16.2` release, please merge it to the `release/16.2` branch, then DM me – I'll be happy to cut a new beta release!

**Comment by diegoreymendez** [comment]
I'm removing the milestone until I resubmit.

**Comment by diegoreymendez** [comment]
@Gio2018 - I fixed these issues:

- Wrong header color in dark mode
- The state of the "My Site" tab is again maintained when switching tabs.

There are some things you noted that are based on the new designs (maybe @osullivanchris can confirm):

- The Me-icon border width was changed from 2.0 points to 0.5 points.
- The Me-icon border-color was changed from `UIColor.white` to `UIColor.separator` in iOS 13+ and `UIColor.gray(.shade20)` in other iOS versions.
- The "My Sites" tab title was changed to "My Site".

That said - all of these UI changes should remain unchanged when the feature flag is OFF.

Can I ask you @Gio2018 to give this another run?  The most important thing is to make sure this PR is not breaking anything especially when the flag is OFF.

I will be addressing @frosty 's feedback about state restoration, and adding back the site picker when the feature flag is on.

**Comment by diegoreymendez** [comment]
To organize my work before my upcoming leave, I'm closing this PR and turning this into the feature branch for Large Titles / White Headers.

The feature branch is named `feature/large-headers`.

**Comment by osullivanchris** [comment]
> There are some things you noted that are based on the new designs (maybe @osullivanchris can confirm):

Those three items all sound correct to me!
